### PR TITLE
BaseRedisService

### PR DESCRIPTION
**Pull request by MarvaZy**

### CODE REVIEW

File: yellowbox/extras/redis/base.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        if self.started:
+            raise RuntimeError("Server already started. Cannot set RDB.")
+        upload_file(self.container, DEFAULT_RDB_PATH, fileobj=redis_file)
+
+    def client_port(self):
+        return get_ports(self.container)[REDIS_DEFAULT_PORT]
+
+    def start(self, retry_spec: Optional[RetrySpec] = None):
+        super().start()
+        self.health()
```
Comment (bentheiii) [inline yellowbox/extras/redis/base.py:40]: health should be retried with retry-specs

File: yellowbox/extras/redis/redis.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        client_cm: ContextManager[Redis] = self.client()
+        with client_cm as client:
+            client.flushall()
+            append_state(client, db_dict)
+
+    def client(self):
+        return redis_client(self)
+
+
+def redis_client(service: RedisService, **kwargs) -> _T:  # type: ignore
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:49]: fix type hint

File: yellowbox/extras/redis/redis.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+from typing import ContextManager, Mapping, Optional, Sequence, Union, TypeVar
+
+from redis import ConnectionError as RedisConnectionError, Redis
+from redis.client import Redis as Redis_Client
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:4]: fix duplicate imports

**Review by bentheiii — CHANGES_REQUESTED** [review]

File: yellowbox/extras/redis/__init__.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from yellowbox.extras.redis.base import REDIS_DEFAULT_PORT, DEFAULT_RDB_PATH, BaseRedisService
+
+try:
+    from yellowbox.extras.redis.redis import RedisService, append_state
+except ImportError:
+    def RedisService():
+        raise ImportError("RedisService can not be imported without redis-py")
+
+    def append_state():
+        raise ImportError("append_state can not be imported without redis-py")
```
Comment (bentheiii) [inline yellowbox/extras/redis/__init__.py:10]:
```suggestion
except ImportError as e:
    def RedisService():
        raise ImportError("RedisService can not be imported without redis-py") from e

    append_state = RedisService
```

File: yellowbox/extras/redis/base.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+from yellowbox.subclasses import SingleContainerService
+
+
+REDIS_DEFAULT_PORT = 6379
+DEFAULT_RDB_PATH = "/data/dump.rdb"
+
+
+class BaseRedisService(SingleContainerService, RunMixin):
+
+    health_exceptions: ClassVar[Catchable] = Exception
```
Comment (bentheiii) [inline yellowbox/extras/redis/base.py:18]: docstring

File: yellowbox/extras/redis/redis.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    health_exceptions: ClassVar[Catchable] = RedisConnectionError
+
+    def health(self, retry_spec: Optional[RetrySpec] = None):
+        client_cm: ContextManager[Redis] = self.client()
+        with client_cm as client:
+            client.ping()
+
+    def reset_state(self):
+        client: Redis
+        with self.client() as client:
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:34]:
```suggestion
        with redis_client(self) as client:
```

File: yellowbox/extras/redis/redis.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        client_cm: ContextManager[Redis] = self.client()
+        with client_cm as client:
+            client.flushall()
+            append_state(client, db_dict)
+
+    def client(self):
+        return redis_client(self)
+
+
+def redis_client(service: RedisService, **kwargs) -> Redis:  # type: ignore
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:47]:
```suggestion
def redis_client(service: BaseRedisService, **kwargs) -> Redis:
```

File: yellowbox/extras/redis/redis.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            client.ping()
+
+    def reset_state(self):
+        client: Redis
+        with self.client() as client:
+            client.flushall()
+
+    def set_state(self, db_dict: RedisState):
+        client_cm: ContextManager[Redis] = self.client()
+        with client_cm as client:
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:39]:
```suggestion
        with redis_client(self) as client:
```

File: yellowbox/retry.py [hunk 9]
```diff
@@ -32,7 +33,7 @@ def __post_init__(self):
             raise ValueError('RetrySpec must have either a timeout or attempts')
 
     def retry(self, func: Callable[[], _T],
-              exceptions: Union[Type[Exception], Tuple[Type[Exception], ...]]) -> _T:
+              exceptions: Catchable) -> _T:
```
Comment (bentheiii) [inline yellowbox/retry.py:36]: combine to one line

File: yellowbox/extras/redis/redis.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            client.rpush(k, *v)
+        elif isinstance(v, Mapping):
+            client.hset(k, mapping=v)  # type: ignore
+        else:
+            client.set(k, v)
+
+
+class RedisService(BaseRedisService):
+
+    health_exceptions: ClassVar[Catchable] = RedisConnectionError
```
Comment (bentheiii) [inline yellowbox/extras/redis/redis.py:25]:
```suggestion
    health_exceptions = RedisConnectionError
```

File: yellowbox/extras/redis/base.py (hunk 5 above)
Comment (bentheiii) [reply yellowbox/extras/redis/base.py:18]: rethinking the default value, check if an empty tuple works, if not, leave it unassigned

**Review by bentheiii — CHANGES_REQUESTED** [review]

File: yellowbox/extras/redis/base.py [hunk 11]
```diff
@@ -1,21 +1,21 @@
...
 
-
 REDIS_DEFAULT_PORT = 6379
 DEFAULT_RDB_PATH = "/data/dump.rdb"
 
 
 class BaseRedisService(SingleContainerService, RunMixin):
 
-    health_exceptions: ClassVar[Catchable] = Exception
+    health_exceptions: ClassVar[Catchable] = tuple()
```
Comment (bentheiii) [inline yellowbox/extras/redis/base.py:17]:
```suggestion
    health_exceptions: ClassVar[Catchable] = ()
```

**Review by bentheiii — APPROVED** [review]

**Review by bharel — CHANGES_REQUESTED** [review]
Design is problematic.

My guess is you've tried to split it so you'll support aioredis and other libraries.
I do not think inheritance for aio-compatibility is the way.

In short:

1. You assume `reset_state()` for example might be aio so it's not in the base class, but you assume `health()` isn't.
2. You don't put `reset_state()` in the base class, which makes the base class not very useful to say the least, as it doesn't contain much more than what `SingleContainerService` already contains, and offers no guarantees about the methods.
3. Importing the modules will cause confusion as you don't know which implementation exists. Usually in those cases, you'll have a variable (`redis.implementations` or `redis.impl`) showing you a tuple of the existing implementations. Trying to use a non-existing implementation will cause a regular AttributeError. This conforms to stdlib and other libraries containing multiple implementations. Having it defined with an error might make things more complicated.
4. You have multiple bugs, for example, the error will actually throw a `TypeError`. These bugs frequently happen when you try to hide a non-existing implementation.

All in all, I think the approach to the solution is the root of the issue. Base classes are not always the answer.

Ben, Aviram and I talked about separating the server from the client many times. I'm still debating on whether it's a good idea or not.

A possible solution: Have a RedisService, and an AsyncRedisService as totally separate classes, no inheritance between them. That's where I'm inclined to atm (Ben opposed a total separation of the client & server before for better convenience, and I don't necessarily disagree).

@aviramha it's the same conversation we had about the azure-storage interface.

**Pull request closed without merging**